src/GraphQL.Tests/Utilities/FederatedSchemaBuilderTests.cs

```diff
@@ -156,5 +157,61 @@ public void input_types_and_types_without_key_directive_are_not_added_to_entitie
             
             Assert.Equal("User", name);
         }
+
+        [Fact]
+        public void resolve_reference_is_not_trying_to_await_for_each_field_individialy_and_plays_well_with_dataloader_issue_1565()
+        {
+            var definitions = @"
+                extend type Query {
+                    user(id: ID!): User
+                }
+                type User @key(fields: ""id"") {
+                    id: ID!
+                    username: String!
+                }
+            ";
+
+            var users = new List<User> {
+                new User { Id = "1", Username = "One" },
+                new User { Id = "2", Username = "Two" },
+            };
+
+
+            var accessor = new DataLoaderContextAccessor
+            {
+                Context = new DataLoaderContext()
+            };
+            var listener = new DataLoaderDocumentListener(accessor);
+
+            Builder.Types.For("User").ResolveReferenceAsync(ctx => {
+                var id = ctx.Arguments["id"].ToString();
+                // return Task.FromResult(users.FirstOrDefault(user => user.Id == id));
+                var loader = accessor.Context.GetOrAddBatchLoader<string, User>("GetAccountByIdAsync", ids => {
+                    var results = users.Where(user => ids.Contains(user.Id));
+                    return Task.FromResult((IDictionary<string, User>)results.ToDictionary(c => c.Id));
+                });
+                return Task.FromResult(loader.LoadAsync(id));
+            });
+
+            var query = @"
+                {
+                    _entities(representations: [{__typename: ""User"", id: ""1"" }, {__typename: ""User"", id: ""2"" }]) {
+                        ... on User {
+                            id
+                            username
+                        }
+                    }
+                }";
+
+            var expected = @"{ ""_entities"": [{ ""__typename"": ""User"", ""id"" : ""1"", ""username"": ""One"" }, { ""__typename"": ""User"", ""id"" : ""2"", ""username"": ""Two"" }] }";
+
+            AssertQuery(_ =>
+            {
+                _.Definitions = definitions;
+                _.Query = query;
+                _.ExpectedResult = expected;
+                _.Listeners.Add(listener);
+            });
+        }
```

Comment:
Maybe a bit of a nit-pick, but it would be nice to see a test that actually was doing async tasks.

Reply:
Actually one already does, by configuring a delay on the mocked results.  See `SimpleDataLoaderTests.Result_Is_Cached`.

I've added a delay to these tests also:
* `BatchDataLoaderTests.Operations_Are_Batched`
* `CollectionBatchLoaderTests.All_Requested_Keys_Should_Be_Cached`